### PR TITLE
New test harness for HTML-based tests

### PR DESCRIPTION
**Pull request by compulim**
## Description

HTML-based test is easier to write and debug than WebDriver tests.

We are not removing WebDriver from the test harness because we still need some help for grabbing image snapshots.

### To-do

- [x] Add code coverage

### Sample HTML-based tests

```js
/**
 * @jest-environment ./__tests__/html/__jest__/WebChatEnvironment.js
 */

describe('offline UI', () => {
  test('should show "unable to connect" UI when credentials are incorrect', () =>
    runHTMLTest('offlineUI.invalidCredentials.html'));
});
```

```html
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <script crossorigin="anonymous" src="https://unpkg.com/@babel/standalone@7/babel.min.js"></script>
    <script crossorigin="anonymous" src="https://unpkg.com/regenerator-runtime/runtime.js"></script>
    <script crossorigin="anonymous" src="https://unpkg.com/react@16.8.6/umd/react.development.js"></script>
    <script crossorigin="anonymous" src="https://unpkg.com/react-dom@16.8.6/umd/react-dom.development.js"></script>
    <script
      crossorigin="anonymous"
      src="https://unpkg.com/react-dom@16.8.6/umd/react-dom-test-utils.development.js"
    ></script>
    <script crossorigin="anonymous" src="/__dist__/testharness.js"></script>
    <link href="/assets/index.css" rel="stylesheet" type="text/css" />
  </head>
  <body>
    <div id="webchat"></div>
    <script crossorigin="anonymous" src="/__dist__/webchat-es5.js"></script>
    <script type="text/babel" data-presets="es2015,stage-3">
      const { conditions, createStore, host, pageObjects, timeouts } = window.WebChatTest;

      (async function() {
        window.WebChat.renderWebChat(
          {
            directLine: window.WebChat.createDirectLine({ token: 'INVALID-TOKEN' }),
            store: createStore(),
            styleOptions: { spinnerAnimationBackgroundImage: 'url(/assets/staticspinner.png)' }
          },
          document.getElementById('webchat')
        );

        await pageObjects.wait(conditions.webChatRendered(), timeouts.ui);
        await pageObjects.wait(conditions.actionDispatched('DIRECT_LINE/CONNECT_REJECTED'), timeouts.directLine);

        await host.snapshot();
        await host.done();
      })().catch(async err => {
        console.error(err);

        await host.error(err);
      });
    </script>
  </body>
</html>
```

## Specific Changes

 

---

-  [x] Testing Added

### CODE REVIEW

File: packages/testharness/src/pageObjects/typeOnSendBox.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import getSendBoxTextBoxElement from '../elements/sendBoxTextBox';
```
Comment (corinagum) [inline packages/testharness/src/pageObjects/typeOnSendBox.js:1]: nit: can we rename this file to `typeInSendBox.js`?

File: __tests__/html/__jest__/mergeCoverageMap.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+const { createCoverageMap } = require('istanbul-lib-coverage');
+
+module.exports = function mergeCoverageMap(...coverageMaps) {
+  const map = createCoverageMap();
+  const addFileCoverage = map.addFileCoverage.bind(map);
+
+  coverageMaps.forEach(coverageMap => Object.values(coverageMap || {}).forEach(addFileCoverage));
+
+  // map.toJSON() do not return a plain object but a serializable object.
```
Comment (corinagum) [inline __tests__/html/__jest__/mergeCoverageMap.js:9]:
```suggestion
  // map.toJSON() does not return a plain object but a serializable object.
```

File: __tests__/html/__jest__/mergeCoverageMap.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  // Jest expect a plain object, thus, we need to stringify/parse to make it a pure JSON.
```
Comment (corinagum) [inline __tests__/html/__jest__/mergeCoverageMap.js:10]:
```suggestion
  // Jest expects a plain object, thus, we need to stringify/parse to make it a pure JSON.
```

File: packages/testharness/src/utils/runAsyncInterval.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+const globalSetTimeout = setTimeout;
```
Comment (corinagum) [inline packages/testharness/src/utils/runAsyncInterval.js:1]: why do we need `globalSetTimeout`?

File: __tests__/setup/setupTestFramework.js [hunk 5]
```diff
@@ -1,27 +1,18 @@
...
 import getPort from 'get-port';
 import handler from 'serve-handler';
 
+import { browserName } from '../constants.json';
 import createPageObjects from './pageObjects/index';
 import marshal from './marshal';
 import retry from './retry';
 import setupTestEnvironment from './setupTestEnvironment';
 
-const BROWSER_NAME = process.env.WEBCHAT_TEST_ENV || 'chrome-docker';
```
Comment (corinagum) [inline __tests__/setup/setupTestFramework.js:14]: Do we really want to remove this?

File: jest.config.js [hunk 6]
```diff
@@ -7,7 +7,8 @@ module.exports = {
     '!<rootDir>/packages/playground/**',
     '!<rootDir>/samples/**'
   ],
-  coverageReporters: ['json', 'lcov', 'text', 'clover', 'cobertura'],
+  coverageReporters: ['json', 'lcov', 'text-summary', 'clover', 'cobertura'],
+  // coverageReporters: ['json', 'lcov', 'text', 'clover', 'cobertura'],
```
Comment (corinagum) [inline jest.config.js:11]: delete?

File: __tests__/html/offlineUI.slowNetwork.reconnect.html [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            })
+          };
+        }
+
+        const clock = lolex.install();
+
+        window.WebChat.renderWebChat(
+          {
+            directLine: createDirectLineForTest({ token: await token.fetchDirectLineToken() }),
+            // directLine: window.WebChat.createDirectLine({ token: await token.fetchDirectLineToken() }),
```
Comment (corinagum) [inline __tests__/html/offlineUI.slowNetwork.reconnect.html:52]: delete?

File: __tests__/html/offlineUI.firstConnect.html [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            })
+          };
+        }
+
+        const clock = lolex.install();
+
+        window.WebChat.renderWebChat(
+          {
+            directLine: createDirectLineForTest({ token: await token.fetchDirectLineToken() }),
+            // directLine: window.WebChat.createDirectLine({ token: await token.fetchDirectLineToken() }),
```
Comment (corinagum) [inline __tests__/html/offlineUI.firstConnect.html:50]: delete?

File: packages/testharness/src/utils/runAsyncInterval.js (hunk 4 above)
Comment (compulim) [reply packages/testharness/src/utils/runAsyncInterval.js:1]: `setTimeout` may be overwritten by lolex (fake timers).

So we save it before it get overwritten.

I added comments in the file for explanation.

File: __tests__/setup/setupTestFramework.js (hunk 5 above)
Comment (compulim) [reply __tests__/setup/setupTestFramework.js:14]: Yes, this is moved to `constants.json` now.

**Pull request merged**